src/commands/sourcemaps/upload.ts

```diff
@@ -113,21 +116,46 @@ export class UploadCommand extends Command {
     return apiConstructor(getBaseIntakeUrl(), this.config.apiKey!)
   }
 
-  private getMatchingSourcemapFiles(): Payload[] {
+  private getMatchingSourcemapFiles = async (): Promise<Payload[]> => {
+    const useGit = this.disableGit === undefined || !this.disableGit
     const sourcemapFiles = glob.sync(buildPath(this.basePath!, '**/*.js.map'))
+    let repositoryData: RepositoryData | undefined
+    if (useGit) {
+      const simpleGit = newSimpleGit()
+      repositoryData = await gitInfos(simpleGit, this.context.stdout, this.repositoryURL)
+    }
 
-    return sourcemapFiles.map((sourcemapPath) => {
-      const minifiedFilePath = getMinifiedFilePath(sourcemapPath)
+    return Promise.all(
+      sourcemapFiles.map(async (sourcemapPath) => {
+        const minifiedFilePath = getMinifiedFilePath(sourcemapPath)
+        let infos: string | undefined
+        if (repositoryData !== undefined && useGit) {
+          const files = await filterTrackedFiles(
+            this.context.stdout,
+            sourcemapPath,
+            this.projectPath,
+            repositoryData.trackedFiles
+          )
+          if (files) {
+            infos = JSON.stringify({
+              files,
+              hash: repositoryData.hash,
+              repository_url: repositoryData.remote,
+            })
+          }
+        }
 
-      return {
-        minifiedFilePath,
-        minifiedUrl: this.getMinifiedURL(minifiedFilePath),
-        projectPath: this.projectPath,
-        service: this.service!,
-        sourcemapPath,
-        version: this.releaseVersion!,
-      }
-    })
+        return {
+          minifiedFilePath,
+          minifiedUrl: this.getMinifiedURL(minifiedFilePath),
+          projectPath: this.projectPath,
+          repository: infos,
+          service: this.service!,
+          sourcemapPath,
+          version: this.releaseVersion!,
+        }
+      })
```

Comment:
In terms of readability, the function's purpose is less clear than before. Ideally the scope of this function should be very high-level:
- look for source maps
- generate something that can be used as a payload for the backend

Now the details about git handling make this function a bit harder to comprehend.

Could we have a function `getGitInfos` that would wraps all those details?
It would be also nice to document this function to explain exactly the behavior implemented (since it has a few edge cases for now, better be explicit about how we are handling this)

Reply:
I reworked the functions in the `upload.ts` file and added comments to make everything more clear. Tell me what you think.